docroot/modules/custom/mass_content/src/Plugin/Field/FieldFormatter/DynamicLinkSeparateFormatter.php

```diff
@@ -78,6 +80,21 @@ public function viewElements(FieldItemListInterface $items, $langcode) {
         '#extra' => $extra,
       ];
 
+      // Add #access and cacheability metadata to remove links to deleted and unpublished content.
+      if ($url->isRouted() && $url->getRouteName() == 'entity.node.canonical') {
+        if ($entity = \Drupal::entityTypeManager()->getStorage(key($url->getRouteParameters()))->load($url->getRouteParameters()['node'])) {
```

Comment:
Oof - this may have a pretty major performance impact, especially for pages like curated list.

Reply:
You quoted the entity load line - is that what you are worried about, or is it the entity access check? I'm not too worried about the entity load. Those should be in entity cache. I dont think the access checks are cached, but they could be since all their contexts, etc are known.

Reply:
An alternative is to queue up deletion of bad entity refs upon entity save but thats gonna cause a lot of invalidation. See my recent comment at https://www.drupal.org/project/drupal/issues/2723323#comment-13956557

Reply:
I'm worried about both loading and access, yeah.  Consider the curated list point for a sec - We're talking about hundreds or thousands of entities that on some of those lists.  Even if all of the entities are in the cache, they're probably not in the static cache, so we've got to pull a lot of data from cache all of a sudden.

I'd be a bigger fan of pruning, which would move the extra work to a separate process we could control.  Understand your point on the invalidation, but this site doesn't actually do that much deletion to begin with, so it might not be so bad.

Reply:
Pruning doesnt actually help us if referenced content gets unpublished or trashbin. We would want the reference to be temporarily removed until the reference is published again. I'll double check with Joe about this requirement.

Reply:
Yeah - I'm not trying to block this. I'm not sure what the requirements or urgency are, and it's totally possible this is no big deal, but I'm concerned that even if this one isn't a big deal, it commits us to a direction that will eventually end up with these kinds of checks being much more widely used in the codebase, so we should be pretty sure it's scalable.

Reply:
FWIW, I do recall quite a bit of work being done at one point to hide unpublished things from showing up. Not sure where you can find that or what form it's in these days, but I do know it was in place.